endpoint: store state in ep_config.json

lmb opened the pull request:
endpoint: remove stat retry on restore

    Commit d81a5cdea5 ("pkg/endpoint: Simplify search for C header file") 
    retained earlier retry behaviour around stat, with a comment saying that we
    can remove the logic if the bug hasn't manifested in a while.

    That was in 2020. The time to remove the check is now.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

endpoint: store state in ep_config.json

    On startup, the agent attemps to restore existing endpoint state from disk.
    It does this by reading the per Endpoint ep_config.h, parsing that file for
    a line containing a magic string and then decoding a base64 encoded JSON
    blob from that. ep_config.h is in turn used by the loader to compile the per
    endpoint BPF programs. In effect, the concern of persisting endpoint state
    is mixed up with how we compile per endpoint programs.

    Instead, write the JSON state into a separate ep_config.json file in
    addition to ep_config.h. The old behaviour is retained so that downgrades do
    not lose endpoint state unnecessarily.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb committed:
endpoint: remove stat retry on restore

Commit d81a5cdea5 ("pkg/endpoint: Simplify search for C header file")
retained earlier retry behaviour around stat, with a comment saying
that we can remove the logic if the bug hasn't manifested in a while.

That was in 2020. The time to remove the check is now.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

squeed reviewed (changes requested):
Can you add some quick upgrade / downgrade test cases?

tommyp1ckles commented on pkg/endpoint/endpoint.go:
```diff
@@ -909,25 +888,19 @@ func FilterEPDir(dirFiles []os.DirEntry) []string {
-	if err := parseBase64ToEndpoint(epSlice[1], &ep); err != nil {
-		return nil, fmt.Errorf("failed to parse restored endpoint: %s", err)
+	if err := ep.UnmarshalJSON(epJSON); err != nil {
+		return nil, fmt.Errorf("error unmarshaling serializableEndpoint from base64 representation: %s", err)
```
nit: error is a bit redundant

lmb replied:
I took that from `parseBase64ToEndpoint`, do you have a preference? Just not wrap the error?

tommyp1ckles commented on pkg/common/const.go:
```diff
@@ -10,6 +10,10 @@ const (
 
+	// EndpointStateFileName is used as the file name for the JSON representation
+	// of endpoint state.
+	EndpointStateFileName = "ep_config.json"
```
Is there any version upgrade considerations with this?

**edit** missed the conversation re: upgrade/dowgrade tests.

vipul-21 commented on pkg/endpoint/restore.go:
```diff
@@ -155,6 +112,55 @@ func ReadEPsFromDirNames(ctx context.Context, owner regeneration.Owner, policyGe
+func findEndpointState(dir string, log *logrus.Entry) ([]byte, error) {
+	state, err := os.ReadFile(filepath.Join(dir, common.EndpointStateFileName))
+	if err == nil {
+		log.Debug("Restoring from JSON file")
```
nit: should this be outside the `if` because we are trying to restore from JSON file by default ? or maybe change it to `restored` instead of restoring.

vipul-21 commented on pkg/endpoint/restore.go:
```diff
@@ -98,40 +77,18 @@ func ReadEPsFromDirNames(ctx context.Context, owner regeneration.Owner, policyGe
-		// This function checks the presence of a bug previously observed: the
-		// file was sometimes only found after the second check.
-		// We can remove this if we haven't seen the issue occur after a while.
-		headerFileExists := func() error {
```
Did we test for this comment ? Is it safe to remove this dual check for file.

lmb replied:
Take a look at the commit message. I think this check was and has been bogus: `stat() followed by open()` is racy and doesn't provide any advantages over simply doing `open()`. The open  call will return `ErrNotExist` or similar just like stat does. This is what the new code does.

lmb committed:
endpoint: store state in ep_config.json

On startup, the agent attempts to restore existing endpoint state
from disk. It does this by reading the per Endpoint ep_config.h,
parsing that file for a line containing a magic string and then
decoding a base64 encoded JSON blob from that. ep_config.h is in
turn used by the loader to compile the per endpoint BPF programs.
In effect, the concern of persisting endpoint state is mixed up
with how we compile per endpoint programs.

Instead, write the JSON state into a separate ep_config.json file
in addition to ep_config.h. The old behaviour is retained so that
downgrades do not lose endpoint state unnecessarily.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>
